src/vaults/SingleStrategyWithdrawal.sol

```diff
@@ -0,0 +1,165 @@
+import {ERC20} from "solmate/src/tokens/ERC20.sol";
+import {SafeTransferLib} from "solmate/src/utils/SafeTransferLib.sol";
+import {FixedPointMathLib} from "solmate/src/utils/FixedPointMathLib.sol";
+
+import {Vault} from "src/vaults/Vault.sol";
+
+contract SingleStrategyWithdrawalEscrow {
+    using SafeTransferLib for ERC20;
+    using FixedPointMathLib for uint256;
+
+    // Epoch counter
+    uint256 public currentEpoch;
+
+    // token paid to user
+    ERC20 public immutable asset;
+
+    // Vault this escrow attached to
+    Vault public immutable vault;
+
+    // map each user with
+    mapping(uint256 => mapping(address => uint256)) public userDebtShare;
+
+    // map per epoch debt share
+    mapping(uint256 => uint256) public epochDebt;
+    // per epoch assets
+    mapping(uint256 => uint256) public epochAsset;
+
+    // last resolved time
+    uint256 public lastResolvedUTCTime;
+
+    constructor(Vault _vault) {
+        asset = ERC20(_vault.asset());
+        vault = _vault;
+    }
+
+    modifier onlyVault() {
+        require(msg.sender == address(vault), "SSWE: must be vault");
+        _;
+    }
+
+    /**
+     * @notice Withdrawal Request event
+     * @param user user address
+     * @param epoch epoch of the request
+     * @param shares withdrawal vault shares
+     * @dev will makes things easy to search for each user withdrawal requests
+     */
+    event WithdrawalRequest(address indexed user, uint256 epoch, uint256 shares);
+
+    /**
+     * @notice Register withdrawal request as debt
+     * @param user user address
+     * @param shares amount of vault shares user requested to withdraw
+     */
+
+    function registerWithdrawalRequest(address user, uint256 shares) external onlyVault {
+        //@ lock user share need to be done from vault, otherwise need approval from user
+        // vault.transferFrom(user, address(this), shares);
+        // register shares of the user
+
+        // check if vault already sent the shares to escrow to lock
+        require((vault.balanceOf(address(this)) - epochDebt[currentEpoch]) == shares, "SSWE: missing shares in escrow.");
+
+        userDebtShare[currentEpoch][user] += shares;
+
+        epochDebt[currentEpoch] += shares;
+
+        emit WithdrawalRequest(user, currentEpoch, shares);
+    }
+
+    /**
+     * @dev this will be a function in vault to check if vault has enough assets to swap locked shares.
+     * @dev will be removed after vault implement this.
+     */
+    function availableToWithdraw(uint256 shares) internal returns (bool) {
+        return shares > 0;
+    }
+
+    /**
+     * @notice resolve the locked shares for current epoch
+     * @dev This function will be triggered after closing a position
+     * @dev will check for available shares to burn
+     * @dev after resolving vault will send the assets to escrow and burn the share
+     */
+    function resolveDebtShares() external onlyVault {
+        uint256 shares = vault.balanceOf(address(this));
```

Comment:
We resolve shares 1 epoch at a time. So we should only be attempting to burn shares for the epoch that was just closed.

Reply:
For single strategy we are not locking any debt. each position is an epoch. So after closing a position we burn all the withdrawal debt.

Reply:
Alright, looks good.